webapp/app.py

```diff
@@ -1,25 +1,108 @@
-from webapp.utils import get_task_status_and_logs
+from webapp.utils import get_task_status_and_logs, parse_oidc_data
 from webapp.utils.aws import ECSClient
 
+logger = logging.getLogger(__name__)
+
+
+class User(UserMixin):
+    """Class representing users of the app.
+
+    This class stores credentials (MIT ID, name, and email)
+    about logged in users.
+    """
+
+    def __init__(
+        self, id: str, name: str, email: str | None = None  # noqa: A002
+    ) -> None:
+        self.id = id
+        self.name = name
+        self.email = email
+
 
 def create_app() -> Flask:
     app = Flask(__name__)
+    app.secret_key = Config().SECRET_KEY
+
+    login_manager = LoginManager()
+    login_manager.init_app(app)
+
+    @login_manager.request_loader
+    def load_user_from_request(request: Request) -> User | None:
+        """Callback to login users using data from request headers.
+
+        For this application, login is the process of identifying the user
+        by decoding data provided in the request headers. This method
+        will:
+
+        * Retrieve OpenID Connect (OIDC) access tokens and user data from request headers.
+           * If access tokens or user data are missing from the request headers,
+             user cannot login.
+        * Stores active access tokens and user data in the Flask session.
+        * Create or refresh a User instance with parsed user data from OIDC.
+
+        An expired access token is marked by a mismatch between the OIDC
+        access token retrieved from the request header and the access token
+        currently stored in the Flask session. If there is a mismatch,
+        the method will parse the updated OIDC data in the request header
+        and refresh the User instance with the updated user data.
+        """
+        if (oidc_access_token := request.headers.get("x-amzn-oidc-accesstoken")) is None:
+            logger.error(
+                "Access token ('x-amzn-oidc-accesstoken') is missing "
+                "from the request headers."
+            )
+            return None
+
+        if (oidc_jwt_data := request.headers.get("x-amzn-oidc-data")) is None:
+            logger.error(
+                "User data ('x-amzn-oidc-data') is missing from the request headers."
+            )
+            return None
+
+        # get user data if new session or refresh user data when access token expires
+        if oidc_access_token != session.get("oidc_access_token"):
+            oidc_data = parse_oidc_data(oidc_jwt_data, options={"verify_exp": False})
+            user = User(
+                id=oidc_data["mit_id"],
+                name=oidc_data["name"],
+                email=oidc_data["preferred_username"],
+            )
+            session.update(
+                {
+                    "oidc_access_token": oidc_access_token,
+                    "user": user.__dict__,
+                }
+            )
+        return User(**session["user"])
```

Comment:
This is going to get a little complicated, but will try and make a case here.

Am I recalling correctly that we had discussed an approach that would only ever instantiate the `User` class once here?  Where instead of storing a `User.__dict__` in the session, we could store the parsed OIDC data and then use that:

1. if token stale or token never set, parse OIDC data from headers
2. store parsed OIDC data -- _not_ the `User` object -- in the session, e.g. `session['oidc_data'] = oidc_data`
3. last, always, instantiate a `User` object by using OIDC data in session:

```python
# last part of function
oidc_data = session['oidc_data']
return User(
  id=oidc_data["mit_id"],
  name=oidc_data["name"],
  email=oidc_data["preferred_username"],
)
```

I would find this more readable, knowing that a `User` object is _only_ instantiated in one way: pulling specific pieces of OIDC data _already_ in the session.  With that cognitive load understood and gone, I could then focus my attention on understanding when and why we might parse OIDC data from headers and set it in the session.

It may seem trivial, but I think this opens up the option of creating a `new()` or `load()` or `from_session_data()` method on the `User` class and using that data, e.g.:

```python
class User(UserMixin):
    # ...
    # ...
    @classmethod
    def from_session_data(cls):
        oidc_data = session.get("oidc_data")
        return cls(
            id=oidc_data["mit_id"],
            name=oidc_data["name"],
            email=oidc_data["preferred_username"],
        )
```

Even if a method like this is not used, I still do think the `@login_manager.request_loader` function having only one line / one way of instantiating a `User` object has value.

Reply:
If we did do something like this, perhaps we could also add a safeguard handler right before we instantiate and return the `User` object, something along the lines of:

```python
# rest of method...

if not session.get('oidc_data'):
    # do something meaningful...
return User.from_session_data()  # or if not the method, just instantiating and returning
```

We _shouldn't_ be in a situation like this where there is not session data to work with... but this ensures if that somehow arises, the app will gracefully handle that.

Reply:
Ah, apologies for missing this! I do remember us talking about replacing the use of `user.__dict__`. I updated the code to [store OIDC data in the session to instantiate the `User`](https://github.com/MITLibraries/alma-sapinvoices-ui/pull/60/commits/d3d80ba1a2083d461ab6358b7f90a8ee758150da#diff-55f12a9f17b4ba61d3e37852150b7d78b8aedbe3629bc4ce8173d4b572297a54R84-R93).

If we did implement a class method like `User.from_session_data`, how do you feel about returning `None` if `oidc_data` is not in the session? 🤔 I think I like it because it would align with `load_user_from_request` returning `None` if it is missing data from the request headers --- which is also worth noting (i.e., because we have these if statements, `session` must have OIDC data before it ever reaches `User` instantiation. 🤔

Reply:
Started a review, but I'll hold off until all changes are finalized. 

I think returning `None` makes a ton of sense.  Like you say, it mirrors the behavior if the headers are missing.  No matter how you slice it: if we can't instantiate a `User`, then return `None` from this method and let the `flask-login` library decide how to handle things.

Good thinking!

Reply:
Agreed ☝️

Reply:
A [`User.from_session_data` class method has been implemented](https://github.com/MITLibraries/alma-sapinvoices-ui/pull/60/files#diff-55f12a9f17b4ba61d3e37852150b7d78b8aedbe3629bc4ce8173d4b572297a54R47-R56). The [`request_loader` callback is updated to use this method](https://github.com/MITLibraries/alma-sapinvoices-ui/pull/60/files#diff-55f12a9f17b4ba61d3e37852150b7d78b8aedbe3629bc4ce8173d4b572297a54R108).

**Note:** I wasn't sure if the idea was to have the condition `if session.get('oidc_data') is None` as part of the `request_loader` callback _or_ part of `User.from_session_data`. 🤔 What I was proposing was the latter -- adding the condition as part of `User.from_session_data`. For what it's worth, I _did_ try initially writing the condition in the callback, but I was getting a lot of linting errors -- the update I went with resolved those errors. Let me know what you think!